File: src/sandbox/fixtures.ts
import type { RawDepositFi } from '../models/deposit';
import { type Rng, createRng, hashSeed } from './rng';
import type { SandboxScenario } from './vua';

/**
 * Deterministic sandbox data. Given a VUA and a run seed, the same accounts and
 * transactions come out every time, so tests and demos are reproducible. Dates are
 * derived from a fixed epoch rather than the wall clock for the same reason.
 */

const SANDBOX_EPOCH_MILLIS = Date.UTC(2026, 5, 1, 9, 30, 0); // 2026-06-01T09:30:00Z
const ONE_DAY_MILLIS = 24 * 60 * 60 * 1000;

interface SandboxFip {
  readonly id: string;
  readonly name: string;
}

const FIPS: readonly SandboxFip[] = [
  { id: 'SBIN', name: 'State Bank of India' },
  { id: 'HDFC', name: 'HDFC Bank' },
];

const CREDIT_NARRATIONS = [
  'Salary credit',
  'UPI/refund',
  'NEFT inward',
  'Interest paid',
  'IMPS from savings',
];
const DEBIT_NARRATIONS = ['UPI/merchant', 'Card purchase', 'Rent payment', 'Electricity bill', 'SIP debit'];
const MODES = ['UPI', 'NEFT', 'IMPS', 'CARD', 'OTHERS'];

/** The result of preparing one FIP's data in the sandbox session. */
export interface FixtureFipData {
  readonly fipId: string;
  readonly fipName: string;
  readonly status: 'DELIVERED' | 'FAILED';
  readonly payload?: RawDepositFi;
}

function paiseToRupeeString(paise: number): string {
  const sign = paise < 0 ? '-' : '';
  const absolute = Math.abs(paise);
  const whole = Math.floor(absolute / 100);
  const fraction = absolute % 100;
  return `${sign}${whole}.${String(fraction).padStart(2, '0')}`;
}

function isoAt(dayOffset: number): string {
  return new Date(SANDBOX_EPOCH_MILLIS - dayOffset * ONE_DAY_MILLIS).toISOString();
}

function buildDepositPayload(fip: SandboxFip, rng: Rng): RawDepositFi {
  const transactionCount = rng.int(8, 14);
  let balancePaise = rng.int(50_000, 400_000) * 100; // opening balance, whole rupees in paise

  const transactions: RawDepositFi['Account']['Transactions']['Transaction'] = [];
  for (let index = transactionCount - 1; index >= 0; index--) {
    const isCredit = rng.next() > 0.45;
    const amountPaise = rng.int(200, 60_000) * 100;
    balancePaise += isCredit ? amountPaise : -amountPaise;
    transactions.push({
      txnId: `${fip.id}-TXN-${String(transactionCount - index).padStart(4, '0')}`,
      type: isCredit ? 'CREDIT' : 'DEBIT',
      amount: paiseToRupeeString(amountPaise),
      currentBalance: paiseToRupeeString(balancePaise),
      valueDate: isoAt(index).slice(0, 10),
      transactionTimestamp: isoAt(index),
      narration: rng.pick(isCredit ? CREDIT_NARRATIONS : DEBIT_NARRATIONS),
      mode: rng.pick(MODES),
      reference: `${fip.id}${rng.int(100000, 999999)}`,
    });
  }

  return {
    Account: {
      maskedAccNumber: `XXXXXX${rng.int(1000, 9999)}`,
      linkedAccRef: `${fip.id}-${rng.int(10000000, 99999999)}`,
      Profile: {
        Holders: {
          Holder: [
            {
              name: rng.pick(['Aarav Sharma', 'Diya Nair', 'Vivaan Rao', 'Ananya Iyer']),
              mobile: '9999999999',
              pan: 'ABCDE1234F',
            },
          ],
        },
      },
      Summary: {
        currentBalance: paiseToRupeeString(balancePaise),
        currency: 'INR',
        balanceDateTime: isoAt(0),
        type: rng.pick(['SAVINGS', 'CURRENT']),
        status: 'ACTIVE',
        branch: rng.pick(['Koramangala', 'Andheri East', 'Salt Lake', 'T Nagar']),
        ifscCode: `${fip.id}0001234`,
      },
      Transactions: { Transaction: transactions },
    },
  };
}

/**
 * Build the per-FIP deposit data for a scenario. Only `healthy` and `partial`
 * produce delivered data here. The earlier failure scenarios (reject, expire,
 * no_accounts, fip_down) are decided by the consent and session state machine
 * before fetch ever reaches fixtures.
 */
export function buildDepositFixtures(
  mobile: string,
  scenario: SandboxScenario,
  seed: number,
): FixtureFipData[] {
  const rng = createRng(hashSeed(`${mobile}:${seed}`));
  const [primaryFip, secondaryFip] = FIPS;
  if (!primaryFip || !secondaryFip) {
    throw new Error('sandbox FIP roster is misconfigured');
  }

  if (scenario === 'partial') {
    return [
      {
        fipId: primaryFip.id,
        fipName: primaryFip.name,
        status: 'DELIVERED',
        payload: buildDepositPayload(primaryFip, rng),
      },
      { fipId: secondaryFip.id, fipName: secondaryFip.name, status: 'FAILED' },
    ];
  }

  return FIPS.map((fip) => ({
    fipId: fip.id,
    fipName: fip.name,
    status: 'DELIVERED' as const,
    payload: buildDepositPayload(fip, rng),
  }));
}

File: python/src/sahaj/sandbox/fixtures.py
"""Deterministic sandbox data. Given a VUA and a run seed, the same accounts and
transactions come out every time, so tests and demos are reproducible. Dates are
derived from a fixed epoch rather than the wall clock for the same reason.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Any, Optional

from .rng import Rng, hash_seed

_SANDBOX_EPOCH = datetime(2026, 6, 1, 9, 30, 0, tzinfo=timezone.utc)


@dataclass(frozen=True)
class _SandboxFip:
    id: str
    name: str


_FIPS: tuple[_SandboxFip, ...] = (
    _SandboxFip("SBIN", "State Bank of India"),
    _SandboxFip("HDFC", "HDFC Bank"),
)

_CREDIT_NARRATIONS = ["Salary credit", "UPI/refund", "NEFT inward", "Interest paid", "IMPS from savings"]
_DEBIT_NARRATIONS = ["UPI/merchant", "Card purchase", "Rent payment", "Electricity bill", "SIP debit"]
_MODES = ["UPI", "NEFT", "IMPS", "CARD", "OTHERS"]


@dataclass(frozen=True)
class FixtureFipData:
    """The result of preparing one FIP's data in the sandbox session."""

    fip_id: str
    fip_name: str
    status: str
    payload: Optional[dict[str, Any]] = None


def _paise_to_rupee_string(paise: int) -> str:
    sign = "-" if paise < 0 else ""
    absolute = abs(paise)
    whole, fraction = divmod(absolute, 100)
    return f"{sign}{whole}.{fraction:02d}"


def _iso_at(day_offset: int) -> str:
    moment = _SANDBOX_EPOCH - timedelta(days=day_offset)
    return moment.isoformat().replace("+00:00", "Z")


def _build_deposit_payload(fip: _SandboxFip, rng: Rng) -> dict[str, Any]:
    transaction_count = rng.int(8, 14)
    balance_paise = rng.int(50_000, 400_000) * 100  # opening balance, whole rupees in paise

    transactions: list[dict[str, Any]] = []
    for index in range(transaction_count - 1, -1, -1):
        is_credit = rng.next() > 0.45
        amount_paise = rng.int(200, 60_000) * 100
        balance_paise += amount_paise if is_credit else -amount_paise
        transactions.append(
            {
                "txnId": f"{fip.id}-TXN-{transaction_count - index:04d}",
                "type": "CREDIT" if is_credit else "DEBIT",
                "amount": _paise_to_rupee_string(amount_paise),
                "currentBalance": _paise_to_rupee_string(balance_paise),
                "valueDate": _iso_at(index)[:10],
                "transactionTimestamp": _iso_at(index),
                "narration": rng.pick(_CREDIT_NARRATIONS if is_credit else _DEBIT_NARRATIONS),
                "mode": rng.pick(_MODES),
                "reference": f"{fip.id}{rng.int(100000, 999999)}",
            }
        )

    return {
        "Account": {
            "maskedAccNumber": f"XXXXXX{rng.int(1000, 9999)}",
            "linkedAccRef": f"{fip.id}-{rng.int(10000000, 99999999)}",
            "Profile": {
                "Holders": {
                    "Holder": [
                        {
                            "name": rng.pick(["Aarav Sharma", "Diya Nair", "Vivaan Rao", "Ananya Iyer"]),
                            "mobile": "9999999999",
                            "pan": "ABCDE1234F",
                        }
                    ]
                }
            },
            "Summary": {
                "currentBalance": _paise_to_rupee_string(balance_paise),
                "currency": "INR",
                "balanceDateTime": _iso_at(0),
                "type": rng.pick(["SAVINGS", "CURRENT"]),
                "status": "ACTIVE",
                "branch": rng.pick(["Koramangala", "Andheri East", "Salt Lake", "T Nagar"]),
                "ifscCode": f"{fip.id}0001234",
            },
            "Transactions": {"Transaction": transactions},
        }
    }


def build_deposit_fixtures(mobile: str, scenario: str, seed: int) -> list[FixtureFipData]:
    """Build the per-FIP deposit data for a scenario.

    Only ``healthy`` and ``partial`` produce delivered data here. The earlier
    failure scenarios are decided by the consent and session state machine before
    fetch ever reaches fixtures.
    """
    rng = Rng(hash_seed(f"{mobile}:{seed}"))
    primary_fip, secondary_fip = _FIPS

    if scenario == "partial":
        return [
            FixtureFipData(primary_fip.id, primary_fip.name, "DELIVERED", _build_deposit_payload(primary_fip, rng)),
            FixtureFipData(secondary_fip.id, secondary_fip.name, "FAILED"),
        ]

    return [
        FixtureFipData(fip.id, fip.name, "DELIVERED", _build_deposit_payload(fip, rng)) for fip in _FIPS
    ]
